## 2
How likely is it that you roll doubles when rolling two dice?


In [1]:
import pandas as pd
import numpy as np



In [2]:
outcomes = [1, 2, 3, 4, 5, 6]

n_simulations = 1_000_000
n_trials = 2

rolls = np.random.choice(outcomes, size=(n_simulations, n_trials))

In [3]:
df=pd.DataFrame()

In [4]:
df = pd.DataFrame(rolls)

In [5]:
df.head()

0  1
0  6  5
1  4  5
2  5  2
3  4  2
4  1  2

In [6]:
df = df.rename(columns={0: "die1", 1: "die2"})

In [7]:
df

die1  die2
0          6     5
1          4     5
2          5     2
3          4     2
4          1     2
...      ...   ...
999995     1     6
999996     4     5
999997     4     4
999998     1     2
999999     5     6

[1000000 rows x 2 columns]

In [8]:
same= df.die1 == df.die2

In [9]:
same.mean()

0.166299

## 2

If you flip 8 coins, what is the probability of getting exactly 3 heads? What is the probability of getting more than 3 heads?

In [10]:
outcomes = [0,1]

n_simulations = 1_000_000
n_trials = 8
flip = np.random.choice(outcomes, size=(n_simulations, n_trials))
flips = pd.DataFrame(flip)

In [11]:
flips

0  1  2  3  4  5  6  7
0       1  0  1  0  1  0  1  1
1       0  1  1  0  0  0  1  0
2       0  0  1  1  1  1  1  0
3       1  1  0  1  0  0  0  0
4       1  1  1  0  1  0  0  1
...    .. .. .. .. .. .. .. ..
999995  0  1  1  1  1  1  1  0
999996  1  1  1  0  1  1  1  0
999997  1  1  1  0  0  0  0  0
999998  0  1  1  0  0  0  1  0
999999  0  1  0  0  1  0  0  1

[1000000 rows x 8 columns]

In [12]:
equals3=flips.sum(axis=1)==3

In [13]:
equals3.mean()

0.219047

In [14]:
greater_than_3 = flips.sum(axis=1)>=3
greater_than_3.mean()

0.856039

## 3
There are approximitely 3 web development cohorts for every 1 data science cohort at Codeup. Assuming that Codeup randomly selects an alumni to put on a billboard, what are the odds that the two billboards I drive past both have data science students on them?

In [15]:
outcomes = [0,1]

n_simulations = 1_000_000
n_trials = 2

students = np.random.choice(outcomes, size=(n_simulations, n_trials), p = [.75,.25])

students=pd.DataFrame(students)

students = students.rename(columns={0: "billboard 1", 1: "billboard 2"})
students

billboard 1  billboard 2
0                 0            1
1                 0            1
2                 0            0
3                 0            0
4                 1            0
...             ...          ...
999995            0            1
999996            0            0
999997            1            0
999998            0            0
999999            0            0

[1000000 rows x 2 columns]

In [16]:
two_data_students = students.sum(axis =1) ==2

In [17]:
two_data_students.mean()

0.062441

## 4
Codeup students buy, on average, 3 poptart packages (+- 1.5) a day from the snack vending machine. If on monday the machine is restocked with 17 poptart packages, how likely is it that I will be able to buy some poptarts on Friday afternoon?


In [18]:
mu, sigma = 3, 1.5
nrows = n_trials = 100_000
ncolms=n_days = 5
poptarts = np.random.normal(mu, sigma, n_trials*n_days).reshape(nrows,ncolms)


In [19]:
poptarts = pd.DataFrame(poptarts)

In [20]:
do_i_poptart=poptarts.sum(axis=1)<17

In [21]:
do_i_poptart.mean()

0.72382


## 5 
Compare Heights

Men have an average height of 178 cm and standard deviation of 8cm.
Women have a mean of 170, sd = 6cm.
Since you have means and standard deviations, you can use np.random.normal to generate observations.
If a man and woman are chosen at random, P(woman taller than man)?


In [25]:
mu, sigma = 178, 8

nrows = n_trials = 100_000

men = np.random.normal(mu, sigma, n_trials)

men=pd.DataFrame(men)
mu2,sigma2= 170, 6

women =  np.random.normal(mu2, sigma2, n_trials)

women = pd.DataFrame(women)


In [26]:

men

0
0      182.223540
1      174.603661
2      177.945668
3      176.168743
4      181.572911
...           ...
99995  188.206733
99996  184.644993
99997  178.346907
99998  176.012127
99999  177.893269

[100000 rows x 1 columns]

In [27]:

women

0
0      175.987214
1      169.706917
2      168.193689
3      161.993096
4      170.285109
...           ...
99995  159.663650
99996  168.388606
99997  168.372596
99998  180.988677
99999  172.856402

[100000 rows x 1 columns]

In [28]:
compare= men < women

In [29]:
compare.mean()

0    0.21029
dtype: float64

## 6

When installing anaconda on a student's computer, there's a 1 in 250 chance that the download is corrupted and the installation fails. What are the odds that after having 50 students download anaconda, no one has an installation issue? 100 students?

What is the probability that we observe an installation issue within the first 150 students that download anaconda?

How likely is it that 450 students all download anaconda without an issue?

In [30]:
install_outcome = [0,1]
n_simulations = 1_000_000
n_trials = 450

downloads = np.random.choice(outcomes, size=(n_simulations, n_trials), p = [.996,.004])
downloads = pd.DataFrame(downloads)
downloads

0    1    2    3    4    5    6    7    8    9    ...  440  441  442  \
0         0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
1         0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
2         0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
3         0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
4         0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
999995    0    0    0    0    0    0    0    0    0    0  ...    0    1    0   
999996    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
999997    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
999998    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
999999    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   

        443  444  445  446  447  448  449  
0         0    0    0    0    0    0    0  
1         0    0    0    0    0    0    0  
2         0    0    0    0    0    0    0  
3         0    0    0    0    0    0    0  
4         0    0    0    0    0    0    0  
...     ...  ...  ...  ...  ...  ...  ...  
999995    0    0    0    0    0    0    0  
999996    0    0    0    0    0    0    0  
999997    0    0    0    0    0    0    0  
999998    0    0    0    0    0    0    0  
999999    0    0    0    0    0    0    0  

[1000000 rows x 450 columns]

In [31]:
failure= downloads.sum(axis=1)<1

In [32]:
failure.mean()

0.164056

50=0.818598

100=0.670997

150=0.451486

450 = 0.164412


## 7 
There's a 70% chance on any given day that there will be at least one food truck at Travis Park. However, you haven't seen a food truck there in 3 days. How unlikely is this?

How likely is it that a food truck will show up sometime this week?



In [33]:
truck_outcome = [0,1]

n_simulations= 100_000

n_trials = 7

food_trucks = np.random.choice(outcomes, size=(n_simulations, n_trials), p = [.7,.3])
food_trucks = pd.DataFrame(food_trucks)
food_trucks=food_trucks.rename(columns={0: "Day 1", 1: "Day 2", 2: "Day 3"})
food_trucks

Day 1  Day 2  Day 3  3  4  5  6
0          0      0      0  1  0  1  0
1          0      0      0  0  0  0  1
2          1      0      1  1  1  1  0
3          1      1      0  0  0  0  1
4          0      1      1  0  0  0  0
...      ...    ...    ... .. .. .. ..
99995      0      0      0  0  1  0  0
99996      1      0      0  0  1  0  0
99997      0      0      0  0  0  0  1
99998      0      0      0  0  0  0  0
99999      0      1      0  0  0  1  0

[100000 rows x 7 columns]

In [35]:
no_food = food_trucks.sum(axis = 1) < 7

In [36]:
no_food.mean()

0.99967

3 days = 0.026719


7 days = 99975


## 7
If 23 people are in the same room, what are the odds that two of them share a birthday? What if it's 20 people? 40?

In [62]:
outcomes = range(1,366)

n_simulations= 1_000_000
n_trials = 20
birthdays = np.random.choice(outcomes, size=(n_simulations, n_trials))
birthdays=pd.DataFrame(birthdays)

In [63]:
birthdays

0    1    2    3    4    5    6    7    8    9    10   11   12   13  \
0       257   31   52  222  164  230   27  365    9   45  114  308  327  204   
1       272   32  211  276   88  233   63  327  364  322   63  364  351  227   
2        82  224   28   52   10  330   74  224  261  335   28  227  228  283   
3       109  156  290  122  317  117  180  312   86  198  330   71  267  335   
4       312  126  123   13   67   15  228  363  202  127  297  232  185  204   
...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
999995   82  309  170   84    6  222  282  112  302   81  206  190  187  142   
999996   87  246  351   64  188   65  326  153  122  144  332  311   50  260   
999997   41   54  351  319   44  266  329   71  343  254  221  164  325   84   
999998   91  340  293  206  279  208  216  105   64    5   77   53  202  218   
999999  326   82   52  221  348  120  129  338   66  202   12  292  244  184   

         14   15   16   17   18   19  
0       269  183  230  278    4   62  
1       127    9  254  240  154  109  
2        77   95   29  215  149  178  
3       312  251   60  339  363  224  
4       129  227  245  312   45  282  
...     ...  ...  ...  ...  ...  ...  
999995  327  145   37  352  143   26  
999996  246   59  188   55  269  222  
999997   31  149  131  294  338   11  
999998  235  325  222   53  318  126  
999999  248  306  296   10  112  133  

[1000000 rows x 20 columns]

In [64]:
no_one_shares_bday = birthdays.apply(lambda row: row.is_unique , axis=1)

In [65]:
1- no_one_shares_bday.mean()

0.411095

23 = 0.506988

20 = 0.411095

40 = 0.890897


# Mage Duel!

Let's use what we've learned to play a mage duel!

Imagine your wizard has 6d4 health points and you have spells that do 6d4 damage. "6d4" means rolling six 4-sided dice and summing the result.

Your opposing mage has 4d6 health points and spells that do 4d6. "4d6" means rolling four six-sided dice and summing the result.

Exercises

Simulate mage duels to answer who is the more powerful mage?

Before running simulations, do you have a hypothesis of which mage will win? Do you have a hunch? Write it down. This is your first exercise.
Simulate 10 mage duels. Is there a clear winner? Run that 10 duel simulation again. Was the answer similar?
Do the results change much at 100 duels?
Now, simulate 10,000 mage duels. Is there a clear winner?

In [83]:
outcomes = [1,2,3,4]
n_simulations= 100_000

n_trials = 6

wizard_1 = np.random.choice(outcomes, size=(n_simulations, n_trials))
wizard_1 = pd.DataFrame(wizard_1)
wizard_power1=wizard_1.sum(axis =1)

In [82]:
outcomes = [1,2,3,4,5,6]
n_simulations= 100_000

n_trials = 4

wizard_2 = np.random.choice(outcomes, size=(n_simulations, n_trials))
wizard_2 = pd.DataFrame(wizard_2)
wizard_power2=wizard_2.sum(axis =1)

In [92]:
battle= pd.concat([wizard_power1,wizard_power2],axis=1)

In [108]:
battle = battle.rename(columns={0: "Mage1", 1: "Mage2"})

In [118]:
battle.where(0 > 1)

ValueError: Array conditional must be same shape as self

In [110]:
outcome = battle.diff(axis=1)

In [111]:
outcome

Mage 1  Mage 2
0         NaN      -1
1         NaN      -2
2         NaN       4
3         NaN       4
4         NaN      -5
...       ...     ...
99995     NaN      10
99996     NaN      -4
99997     NaN      -2
99998     NaN      -8
99999     NaN      -5

[100000 rows x 2 columns]

In [114]:
outcome_duel = outcome > 0

In [115]:
outcome_duel.mean()

Mage 1    0.00000
Mage 2    0.36914
dtype: float64